dataRetriever/dataPool/headersCache/headersPool_test.go

```diff
@@ -0,0 +1,342 @@
+	assert.Equal(t, &hdrs[1], hdr1)
+
+	hdr2, err := hdrsCacher.GetHeaderByHash(hdrsHashes[2])
+	assert.Nil(t, err)
+	assert.Equal(t, &hdrs[2], hdr2)
+}
+
+func TestHeadersCacher_AddHeadersWithSameNonceShouldBeRemovedAtEviction(t *testing.T) {
+	t.Parallel()
+
+	shardId := uint32(0)
+	cacheSize := 2
+
+	hash1, hash2, hash3 := []byte("hash1"), []byte("hash2"), []byte("hash3")
+	hdr1, hdr2, hdr3 := &block.Header{Nonce: 0}, &block.Header{Nonce: 0}, &block.Header{Nonce: 1}
+
+	hdrsCacher, _ := headersCache.NewHeadersCacher(cacheSize, 1)
+	hdrsCacher.Add(hash1, hdr1)
+	hdrsCacher.Add(hash2, hdr2)
+	hdrsCacher.Add(hash3, hdr3)
+
+	assert.Equal(t, 1, hdrsCacher.GetNumHeadersFromCacheShard(shardId))
+
+	hdr, err := hdrsCacher.GetHeaderByHash(hash3)
+	assert.Nil(t, err)
+	assert.Equal(t, hdr3, hdr)
+}
+
+func TestHeadersCacher_AddALotOfHeadersAndCheckEviction(t *testing.T) {
+	t.Parallel()
+
+	cacheSize := 100
+	numHeaders := 500
+	shardId := uint32(0)
+	hdrs, hdrsHash := createASliceOfHeaders(numHeaders, shardId)
+	hdrsCacher, _ := headersCache.NewHeadersCacher(cacheSize, 50)
+
+	var waitgroup sync.WaitGroup
+	for i := 0; i < numHeaders; i++ {
+		waitgroup.Add(1)
+		go func(index int) {
+			hdrsCacher.Add(hdrsHash[index], &hdrs[index])
+			waitgroup.Done()
+		}(i)
+	}
+
+	waitgroup.Wait()
+	assert.Equal(t, 100, hdrsCacher.GetNumHeadersFromCacheShard(shardId))
+}
+
+func TestHeadersCacher_BigCacheALotOfHeadersShouldWork(t *testing.T) {
+	t.Parallel()
+
+	cacheSize := 100000
+	numHeadersToGenerate := cacheSize
+	shardId := uint32(0)
+
+	hdrs, hdrsHash := createASliceOfHeaders(numHeadersToGenerate, shardId)
+	hdrsCacher, _ := headersCache.NewHeadersCacher(cacheSize, 50)
+
+	start := time.Now()
+	for i := 0; i < numHeadersToGenerate; i++ {
+		hdrsCacher.Add(hdrsHash[i], &hdrs[i])
+	}
+	elapsed := time.Since(start)
+	fmt.Printf("insert %d took %s \n", numHeadersToGenerate, elapsed)
+
+	start = time.Now()
+	hdr, _ := hdrsCacher.GetHeaderByHash(hdrsHash[100])
+	elapsed = time.Since(start)
+	assert.Equal(t, &hdrs[100], hdr)
+	fmt.Printf("get header by hash took %s \n", elapsed)
+
+	start = time.Now()
+	d, _, _ := hdrsCacher.GetHeaderByNonceAndShardId(uint64(100), shardId)
+	elapsed = time.Since(start)
+	fmt.Printf("get header by shard id and nonce took %s \n", elapsed)
+	assert.Equal(t, &hdrs[100], d[0])
+
+	start = time.Now()
+	hdrsCacher.RemoveHeaderByNonceAndShardId(uint64(500), shardId)
+	elapsed = time.Since(start)
+	fmt.Printf("remove header by shard id and nonce took %s \n", elapsed)
+
+	hdr, err := hdrsCacher.GetHeaderByHash(hdrsHash[500])
+	assert.Error(t, headersCache.ErrHeaderNotFound, err)
+
+	start = time.Now()
+	hdrsCacher.RemoveHeaderByHash(hdrsHash[2012])
+	elapsed = time.Since(start)
+	fmt.Printf("remove header by hash took %s \n", elapsed)
+
+	hdr, err = hdrsCacher.GetHeaderByHash(hdrsHash[2012])
+	assert.Error(t, headersCache.ErrHeaderNotFound, err)
+}
+
+func TestHeadersCacher_AddHeadersWithDifferentShardIdOnMultipleGoroutines(t *testing.T) {
```

Comment:
This seems to be the only test with more than one shards? If this is correct, could you please add more, but simpler, tests for Add / Remove for more shards?

Reply:
done.